prov/sockets/src/sock_progress.c

```diff
@@ -738,14 +739,29 @@ static int sock_pe_process_rx_read(struct sock_pe *pe,
 		return 0;
 	len += entry_len;
 
+	if (rx_ctx->domain->attr.mr_mode & FI_MR_RAW) {
+		size_t total_raw_size = rx_ctx->domain->attr.mr_key_size *
+				pe_entry->msg_hdr.dest_iov_len;
+		raw_keys = malloc(total_raw_size);
+		if (!raw_keys)
+			return -FI_ENOMEM;
+		if (sock_pe_recv_field(pe_entry, raw_keys, total_raw_size, len))
+			return 0;
+		len += total_raw_size;
+	}
```

Comment:
The changes in this function are identical to those below.  Please move common code into a new function.